ej_cli/sis/sis.py

```diff
@@ -0,0 +1,245 @@
+from email import header
+import os
+import re
+import json
+from typing import Optional
+import requests
+from bs4 import BeautifulSoup
+from tabulate import tabulate
+from .regex import matches
+
+
+BASE_DIR = os.path.dirname(os.path.abspath(__file__))
+PARENT_DIR = os.path.dirname(BASE_DIR)
+
+
+class SIS:
+    def __init__(self, id: str, pwd: str, save_credentials: bool = True, new_pwd: bool = False) -> None:
+        self.id = id
+        self.pwd = pwd
+        self.save_credentials = save_credentials
+        self.__data = {}
+        self.__session = None
+        self.__authenticated = False
+
+        if new_pwd and os.path.exists(
+            os.path.join(PARENT_DIR, f"saved/{self.id}-credentials.json")
+        ):
+            os.remove(os.path.join(
+                PARENT_DIR, f"saved/{self.id}-credentials.json"))
+
+        if self.save_credentials and not os.path.exists(
+            os.path.join(PARENT_DIR, f"saved/{self.id}-credentials.json")
+        ):
+            with open(
+                os.path.join(
+                    PARENT_DIR, f"saved/{self.id}-credentials.json"
+                ), "w"
+            ) as f:
+                json.dump({"id": id, "pwd": pwd}, f, indent=4)
+
+    def open_session(self) -> requests.Session:
+        if not self.__session:
+            self.__session = requests.Session()
+
+        return self.__session
+
+    def close_session(self, req: requests.Session) -> None:
+        req.close()
+        self.session = None
+
+    def authenticate(self) -> None:
+        if self.__authenticated:
+            return
+
+        if not self.__session:
+            self.open_session()
+
+        login_page = self.req("https://sis.ejust.edu.eg/Default.aspx")
+        soup = BeautifulSoup(login_page.content, features="html5lib")
+        VIEWSTATE = soup.find(id="__VIEWSTATE")["value"]
+        VIEWSTATEGENERATOR = soup.find(id="__VIEWSTATEGENERATOR")["value"]
+        EVENTVALIDATION = soup.find(id="__EVENTVALIDATION")["value"]
+        login_data = {
+            "__EVENTTARGET": "",
+            "__EVENTARGUMENT": "",
+            "__VIEWSTATE": VIEWSTATE,
+            "__VIEWSTATEGENERATOR": VIEWSTATEGENERATOR,
+            "__EVENTVALIDATION": EVENTVALIDATION,
+            "txtUsername": self.id,
+            "txtPassword": self.pwd,
+            "btnEnter": "Log In"
+        }
+        self.req(
+            "https://sis.ejust.edu.eg/Default.aspx", method="POST", data=login_data
```

Comment:
extract URL to constant b2a

Reply:
:+1: 
Also, I will add new issues if you want to contribute :heart: